runtime-sdk/src/storage/confidential.rs

```diff
@@ -0,0 +1,422 @@
+        common::crypto::{
+            hash::Hash,
+            mrae::deoxysii::{self, NONCE_SIZE},
+        },
+        storage::mkvs,
+    },
+    storage::Store,
+};
+
+type Nonce = [u8; NONCE_SIZE];
+type Kdf = Hmac<Sha512Trunc256>;
+
+pub struct EncryptionKey([u8; KEY_SIZE]);
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("corrupt key")]
+    CorruptKey,
+
+    #[error("corrupt value")]
+    CorruptValue,
+
+    #[error("decryption failure: {0}")]
+    DecryptionFailure(anyhow::Error),
+}
+
+/// A key-value store that encrypts all content with DeoxysII.
+pub struct ConfidentialStore<S: Store> {
+    inner: S,
+    deoxys: deoxysii::DeoxysII,
+    base_value_prefix: Vec<u8>,
+    nonce_counter: usize,
+    nonce_key: Zeroizing<Vec<u8>>,
+}
+
+impl<S: Store> ConfidentialStore<S> {
+    /// Create a new confidential store with the given keypair.
+    pub fn new_with_key(inner: S, key: [u8; KEY_SIZE], value_context: &[&[u8]]) -> Self {
+        let actual_key = Zeroizing::new(key);
+        let mut kdf = Kdf::new_from_slice(b"oasis-runtime-sdk/confidential store")
+            .expect("Hmac::new_from_slice");
+        kdf.update(&key);
+        let mut derived = kdf.finalize().into_bytes();
+        let derived = Zeroizing::new(derived.iter_mut());
+        ConfidentialStore {
+            inner,
+            deoxys: deoxysii::DeoxysII::new(&actual_key),
+            base_value_prefix: value_context.concat(),
+            nonce_counter: 0,
+            nonce_key: Zeroizing::new(derived.as_slice().to_vec()),
+        }
+    }
+
+    fn pack_nonce_slice(&self, nonce: &Nonce, slice: &[u8]) -> Vec<u8> {
+        let mut ret = Vec::with_capacity(nonce.len() + slice.len());
+        ret.extend_from_slice(nonce);
+        ret.extend_from_slice(slice);
+        ret
+    }
+
+    fn unpack_nonce_slice<'a>(&self, packed: &'a [u8]) -> Option<(&'a Nonce, &'a [u8])> {
+        if packed.len() <= NONCE_SIZE {
+            return None;
+        }
+        let nonce_ref: &'a Nonce = packed[..NONCE_SIZE]
+            .try_into()
+            .expect("nonce size mismatch");
+        Some((nonce_ref, &packed[NONCE_SIZE..]))
+    }
+
+    fn make_key(&self, plain_key: &[u8]) -> (Nonce, Vec<u8>) {
+        let mut nonce = [0u8; NONCE_SIZE];
+
+        let mut nonce_src = self.nonce_key.clone();
+        nonce_src.extend_from_slice(plain_key);
+
+        let hash = Hash::digest_bytes(&nonce_src);
+        nonce.copy_from_slice(hash.truncated(NONCE_SIZE));
+        let enc_key = self.deoxys.seal(&nonce, plain_key.to_vec(), vec![]);
+        let key = self.pack_nonce_slice(&nonce, &enc_key);
+        (nonce, key)
+    }
+
+    fn make_value(&mut self, plain_value: &[u8]) -> (Nonce, Vec<u8>) {
+        let mut nonce = [0u8; NONCE_SIZE];
+
+        self.nonce_counter += 1;
+        let hash = Hash::digest_bytes_list(&[
+            self.base_value_prefix.as_slice(),
+            self.nonce_counter.to_le_bytes().as_slice(),
+            plain_value,
```

Comment:
No need to put value here given that the above should already be fairly unique. If you put the value here this would again require the "nonce key treatment".